cpp/src/cider/type/plan/BinaryExpr.cpp

```diff
@@ -69,23 +71,66 @@ JITExprValue& BinOper::codegen(CodegenContext& context) {
     if (get_optype() == kBW_EQ or get_optype() == kBW_NE) {
       return codegenFixedSizeDistinctFrom(func, lhs_val, rhs_val);
     }
-    JITValuePointer null = func.createVariable(JITTypeTag::BOOL, "null_val");
-    null = lhs_val.getNull() || rhs_val.getNull();
 
     const auto optype = get_optype();
     if (IS_ARITHMETIC(optype)) {
+      auto null = lhs_val.getNull() || rhs_val.getNull();
       return codegenFixedSizeColArithFun(
           context, null, lhs_val.getValue(), rhs_val.getValue());
     } else if (IS_COMPARISON(optype)) {
+      auto null = lhs_val.getNull() || rhs_val.getNull();
       return codegenFixedSizeColCmpFun(null, lhs_val.getValue(), rhs_val.getValue());
     } else if (IS_LOGIC(optype)) {
-      return codegenFixedSizeLogicalFun(func, null, lhs_val, rhs_val);
+      return codegenFixedSizeLogicalFun(context, func, lhs_val, rhs_val);
     }
   }
   UNREACHABLE();
   return expr_var_;
 }
 
+JITExprValue& BinOper::codegenNull(CodegenContext& context) {
+  // JITFunction& func = *context.getJITFunction();
+  // FIXME: how to use cache for not first call
+  // cann't use cache in the first call,
+  // cause codegen() generated null operation need to be updated
+  // if (auto& expr_var = get_expr_value()) {
+  //   return expr_var;
+  // }
+
+  auto lhs = const_cast<Analyzer::Expr*>(get_left_operand());
+  auto rhs = const_cast<Analyzer::Expr*>(get_right_operand());
+
+  if (is_unnest(lhs) || is_unnest(rhs)) {
+    CIDER_THROW(CiderCompileException, "Unnest not supported in comparisons");
+  }
+
+  const auto& lhs_ti = lhs->get_type_info();
+  const auto& rhs_ti = rhs->get_type_info();
+  if (lhs_ti.is_string()) {
+    CHECK(rhs_ti.is_string());
+  } else {
+    CHECK_EQ(lhs_ti.get_type(), rhs_ti.get_type());
+  }
+  if (lhs_ti.is_decimal() || lhs_ti.is_timeinterval()) {
+    CIDER_THROW(CiderCompileException,
+                "Decimal and TimeInterval are not supported in arithmetic codegen now.");
+  }
+  if (lhs_ti.is_string()) {
+    // string binops, should only be comparisons
+    // const auto optype = get_optype();
+    UNIMPLEMENTED();
+  } else {
+    // primitive type binops
+    JITExprValueAdaptor lhs_val(lhs->codegenNull(context));
+    JITExprValueAdaptor rhs_val(rhs->codegenNull(context));
+
+    // FIXME: propagate null, don't support kleene logic operation
+    return set_expr_null(lhs_val.getNull() || rhs_val.getNull());
+  }
+  UNREACHABLE();
+  return expr_var_;
```

Comment:
can we move this return stmt and insert after line [121](https://github.com/intel/BDTK/pull/331/files#diff-69477261c52d5c47959aa986fd21a9fc5608440612e33637aa7bcf6747c86981R121)

Reply:
yes, we can, if we accept this as the final strategy, I'll polish this code.